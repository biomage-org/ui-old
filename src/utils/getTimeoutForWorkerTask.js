import getNumberOfCellsInGrouping from 'redux/selectors/getNumberOfCellsInGrouping';
import { getCellSetsHierarchyByKeys } from 'redux/selectors';

// Timeouts calculated in https://docs.google.com/document/d/1vim9t9lWMLW8wALeJvDeYnofQa9tj9zPU3i1SOfMilM/edit
const getTimeoutForWorkerTaskUncapped = (state, taskName, options) => {
  // Get filtered nCells for more accurate timeout//
  // if louvain is not calculated (unlikely) get all nCells
  const nCells = getNumberOfCellsInGrouping('louvain', state) ?? getNumberOfCellsInGrouping('sample', state);
  const nClusters = getCellSetsHierarchyByKeys(['louvain'])(state)[0]?.children.length ?? 1;

  switch (taskName) {
    case 'GetEmbedding': {
      const { type } = options;

      // Tsne is slower than tsne, so we give a bigger timeout to tsne
      if (type === 'umap') return 0.002 * nCells + 60;
      if (type === 'tsne') return 0.02 * nCells + 60;

      throw new Error('GetEmbedding type isn\'t specified');
    }
    case 'ClusterCells':
    case 'MarkerHeatmap': {
      return 0.002 * nCells + 60;
    }
    case 'DifferentialExpression': {
      return 180;
    }
    case 'TrajectoryAnalysisStartingNodes': {
      // Number of clusters is inversely proportional to running time
      // Larger custers produce more possible trajectories
      return ((0.3 * nCells) / nClusters) + 60;
    }
    case 'TrajectoryAnalysisPseudotime': {
      return ((0.6 * nCells) / nClusters) + 60;
    }
    case 'ListGenes':
    case 'GeneExpression':
    case 'GetMitochondrialContent':
    case 'GetDoubletScore':
    case 'GetNormalizedExpression':
    case 'GetNUmis':
    case 'GetNGenes':
    case 'PlotData': {
      return 60;
    }
    default: {
      throw new Error('Task doesn\'t exist');
    }
  }
};

const getTimeoutForWorkerTask = (state, taskName, options) => (
  Math.max(getTimeoutForWorkerTaskUncapped(state, taskName, options), 60)
);

export default getTimeoutForWorkerTask;
