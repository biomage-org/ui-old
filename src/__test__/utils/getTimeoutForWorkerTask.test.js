import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';

jest.mock('redux/selectors/getNumberOfCellsInGrouping', () => ({
  __esModule: true, // this property makes it work
  default: () => 20,
}));

jest.mock('redux/selectors', () => ({
  getCellSetsHierarchyByKeys: () => () => 1,
}));

describe('getTimeoutForWorkerTask', () => {
  const stateMock = {};

  it('works correctly for GetEmbedding with umap', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'GetEmbedding', { type: 'umap' });
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for GetEmbedding with tsne', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'GetEmbedding', { type: 'tsne' });
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for ClusterCells', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'ClusterCells');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for MarkerHeatmap', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'MarkerHeatmap');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for DifferentialExpression', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'DifferentialExpression');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for ListGenes', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'ListGenes');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for GeneExpression', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'GeneExpression');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for GetMitochondrialContent', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'GetMitochondrialContent');
    expect(timeout).toMatchSnapshot();
  });

  it('works correctly for GetDoubletScore', () => {
    const timeout = getTimeoutForWorkerTask(stateMock, 'GetDoubletScore');
    expect(timeout).toMatchSnapshot();
  });

  it('throws for invalid taskName', () => {
    expect(() => getTimeoutForWorkerTask(stateMock, 'InvalidTaskName')).toThrowError('Task doesn\'t exist');
  });
});
