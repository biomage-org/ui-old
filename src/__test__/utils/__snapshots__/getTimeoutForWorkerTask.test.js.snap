// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTimeoutForWorkerTask works correctly for ClusterCells 1`] = `180.04`;

exports[`getTimeoutForWorkerTask works correctly for DifferentialExpression 1`] = `180`;

exports[`getTimeoutForWorkerTask works correctly for GeneExpression 1`] = `180`;

exports[`getTimeoutForWorkerTask works correctly for GetDoubletScore 1`] = `180`;

exports[`getTimeoutForWorkerTask works correctly for GetEmbedding with tsne 1`] = `900`;

exports[`getTimeoutForWorkerTask works correctly for GetEmbedding with umap 1`] = `900`;

exports[`getTimeoutForWorkerTask works correctly for GetMitochondrialContent 1`] = `180`;

exports[`getTimeoutForWorkerTask works correctly for ListGenes 1`] = `900`;

exports[`getTimeoutForWorkerTask works correctly for MarkerHeatmap 1`] = `900`;
